File: backend/service-carts/src/main/java/com/onlinestore/servicecarts/controller/CartController.java
package com.onlinestore.servicecarts.controller;

import com.onlinestore.servicecarts.dto.CartDTO;
import com.onlinestore.servicecarts.service.CartService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;


@RestController
@RequestMapping("/api/v1/carts")
@CrossOrigin("http://localhost:4000")
public class CartController {
    @Autowired
    private CartService cartService;

    @GetMapping("/user/{userId}")
    public ResponseEntity<Page<CartDTO>> getCartsByUserId(
            @PathVariable String userId,
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "3") int size) {
        PageRequest pageRequest = PageRequest.of(page, size);
        Page<CartDTO> carts = cartService.getCartsByUserId(userId, pageRequest);
        return ResponseEntity.ok(carts);
    }

    @GetMapping("/{cartId}")
    public ResponseEntity<CartDTO> getCartById(@PathVariable String cartId) {
        CartDTO cart = cartService.getCartById(cartId);
        return ResponseEntity.ok(cart);
    }

    @PostMapping
    public ResponseEntity<CartDTO> createCart(@RequestBody CartDTO cartDto) {
        CartDTO newCart = cartService.createCart(cartDto);
        return ResponseEntity.ok(newCart);
    }

    @DeleteMapping("/{cartId}")
    public ResponseEntity<Void> deleteCart(@PathVariable String cartId) {
        cartService.deleteCart(cartId);
        return ResponseEntity.noContent().build();
    }

    @PutMapping("/{cartId}")
    public ResponseEntity<CartDTO> updateCart(@PathVariable String cartId, @RequestBody CartDTO cartDto) {
        CartDTO updatedCart = cartService.updateCart(cartId, cartDto);
        return ResponseEntity.ok(updatedCart);
    }
}

File: bff/src/controllers/cartController.js
const cartService = require("../services/cartService");

const getCartsByUserId = async (req, res) => {
  try {
    const { userId } = req.params;
    const { page = 0, size = 3 } = req.query;
    const carts = await cartService.getCartsByUserId(userId, page, size);
    res.json(carts);
  } catch (error) {
    console.error("Error fetching carts:", error);
    res.status(500).json({ message: "Error fetching carts", error });
  }
};

const getCartByCartId = async (req, res) => {
  try {
    const { cartId } = req.params;
    const cart = await cartService.getCartByCartId(cartId);
    res.json(cart);
  } catch (error) {
    res.status(500).json({ message: "Error fetching cart", error });
  }
};

const createCart = async (req, res) => {
  try {
    const cartDto = req.body;
    const newCart = await cartService.createCart(cartDto);
    res.json(newCart);
  } catch (error) {
    res.status(500).json({ message: "Error creating cart", error });
  }
};

const deleteCart = async (req, res) => {
  try {
    const { cartId } = req.params;
    await cartService.deleteCart(cartId);
    res.status(204).send();
  } catch (error) {
    res.status(500).json({ message: "Error deleting cart", error });
  }
};

const updateCart = async (req, res) => {
  try {
    const { cartId } = req.params;
    const cartDto = req.body;
    const updatedCart = await cartService.updateCart(cartId, cartDto);
    res.json(updatedCart);
  } catch (error) {
    res.status(500).json({ message: "Error updating cart", error });
  }
};

module.exports = {
  getCartsByUserId,
  getCartByCartId,
  createCart,
  deleteCart,
  updateCart,
};
